### PR TITLE
OSPP: Development of Federated Incremental Learning for Label Scarcity: Base on KubeEdge-Ianvs

### PR DESCRIPTION
**Pull request by Yoda-wu**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR add two main feature: 
- federated learning paradigm
- federated class incremental learning paradigm

Besides this PR also add the example of these two paradigm.

**Which issue(s) this PR fixes**:
related issue: https://github.com/kubeedge/ianvs/issues/97

Fixes #

### CODE REVIEW

**Review by MooreZheng — CHANGES_REQUESTED** [review]
There are three CI issues that remain to be resolved, mainly for Pylint, see https://github.com/kubeedge/ianvs/actions/runs/11268242806

**Comment by Yoda-wu** [comment]
> There are three CI issues that remain to be resolved, mainly for Pylint, see https://github.com/kubeedge/ianvs/actions/runs/11268242806

Thanks for comment!, but i think the linked action is not my pr's action.
I have some question about Pylint, In Pylint(3.9) it got error:
```
Run pylint '/home/runner/work/ianvs/ianvs/core'
************* Module core.testenvmanager.dataset.dataset
core/testenvmanager/dataset/dataset.py:124:4: R091[7](https://github.com/kubeedge/ianvs/actions/runs/11320404587/job/31477907116#step:5:8): Too many positional arguments (8/5) (too-many-positional-arguments)
core/testenvmanager/dataset/dataset.py:237:4: R0917: Too many positional arguments (6/5) (too-many-positional-arguments)
core/testenvmanager/dataset/dataset.py:244:4: R0[9](https://github.com/kubeedge/ianvs/actions/runs/11320404587/job/31477907116#step:5:10)17: Too many positional arguments (7/5) (too-many-positional-arguments)
core/testenvmanager/dataset/dataset.py:291:4: R0917: Too many positional arguments (7/5) (too-many-positional-arguments)
core/testenvmanager/dataset/dataset.py:349:4: R0917: Too many positional arguments (7/5) (too-many-positional-arguments)
core/testenvmanager/dataset/dataset.py:422:4: R0917: Too many positional arguments (7/5) (too-many-positional-arguments)
core/testenvmanager/dataset/dataset.py:492:4: R0917: Too many positional arguments (6/5) (too-many-positional-arguments)
************* Module core.testcasecontroller.algorithm.paradigm.singletask_learning.singletask_learning_active_boost
core/testcasecontroller/algorithm/paradigm/singletask_learning/singletask_learning_active_boost.py:66:4: R0917: Too many positional arguments (7/5) (too-many-positional-arguments)
```
which is not my pr's feature. I try to add  disabel too-many-positional-arguments and it turn outs that Pylint(3.7,3.8) can't recognize this argument.
```
Run pylint '/home/runner/work/ianvs/ianvs/core'
************* Module core.testenvmanager.dataset.dataset
core/testenvmanager/dataset/dataset.py:119:0: W0012: Unknown option value for 'disable', expected a valid pylint message and got 'too-many-positional-arguments' (unknown-option-value)
core/testenvmanager/dataset/dataset.py:20[7](https://github.com/kubeedge/ianvs/actions/runs/11137579503/job/31477583235?pr=143#step:5:8):0: W0012: Unknown option value for 'disable', expected a valid pylint message and got 'too-many-positional-arguments' (unknown-option-value)
core/testenvmanager/dataset/dataset.py:215:0: W0012: Unknown option value for 'disable', expected a valid pylint message and got 'too-many-positional-arguments' (unknown-option-value)
core/testenvmanager/dataset/dataset.py:249:0: W0012: Unknown option value for 'disable', expected a valid pylint message and got 'too-many-positional-arguments' (unknown-option-value)
core/testenvmanager/dataset/dataset.py:2[8](https://github.com/kubeedge/ianvs/actions/runs/11137579503/job/31477583235?pr=143#step:5:9)8:0: W0012: Unknown option value for 'disable', expected a valid pylint message and got 'too-many-positional-arguments' (unknown-option-value)
core/testenvmanager/dataset/dataset.py:334:0: W0012: Unknown option value for 'disable', expected a valid pylint message and got 'too-many-positional-arguments' (unknown-option-value)
```

File: examples/cifar100/fci_ssl/fed_ci_match/algorithm/agumentation.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        return tf.image.random_crop(y, tf.shape(x))
+
+    def augment_shift_mnist(self, x, w):
+        y = tf.pad(x, [[0] * 2, [w] * 2, [w] * 2], mode="REFLECT")
+        return tf.image.random_crop(y, tf.shape(x))
+
+    def augment_for_cifar(self, images: np.ndarray):
+        return self.augment_shift(self.augment_mirror(images), 4)
+
+    def augment_for_svhn(self, images: np.ndarray):
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/fed_ci_match/algorithm/agumentation.py:198]: Unused dataset augmentation code should be removed

File: examples/cifar100/fci_ssl/fed_ci_match/algorithm/agumentation.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        return self.augment_shift_mnist(self.augment_mirror(images), 4)
+
+
+class RandAugment(Base_Augment):
+    def __init__(self, dataset_name: str) -> None:
+        super().__init__(dataset_name)
+        self.rand_augment = Rand_Augment()
+        if self.dataset_name in ["cifar10", "cifar100", "svhn"]:
+            self.input_shape = (32, 32, 3)
+        elif self.dataset_name == "stl10":
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/fed_ci_match/algorithm/agumentation.py:238]: Unused dataset augmentation code should be removed

File: core/testcasecontroller/algorithm/paradigm/federated_learning/federated_learning.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        run the test flow of incremental learning paradigm.
+
+        Returns
+        ------
+        test result: numpy.ndarray
+        system metric info: dict
+            information needed to compute system metrics.
+        """
+        # init client wait for connection
+        # self.init_client()
```
Comment (chenhaorui0768) [inline core/testcasecontroller/algorithm/paradigm/federated_learning/federated_learning.py:92]: Annotated code should be removed

File: examples/cifar100/fci_ssl/fedavg/algorithm/aggregation.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        -------
+        update_weights : Array-like
+            final weights use to update model layer
+        """
+
+        print("aggregation....")
+        if not len(clients):
+            return self.weights
+        self.total_size = sum([c.num_samples for c in clients])
+        # print(next(iter(clients)).weights)
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/fedavg/algorithm/aggregation.py:39]: Annotated code should be removed

File: examples/cifar100/fci_ssl/fedavg/algorithm/basemodel.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            x = np.load(data)
+            logging.info(f"predicting {x.shape}")
+            mean = np.array((0.5071, 0.4867, 0.4408), np.float32).reshape(1, 1, -1)
+            std = np.array((0.2675, 0.2565, 0.2761), np.float32).reshape(1, 1, -1)
+            x = (tf.cast(x, dtype=tf.float32) / 255.0 - mean) / std
+            pred = self.classifier(self.fe(x, training=False))
+            prob = tf.nn.softmax(pred, axis=1)
+            pred = tf.argmax(prob, axis=1)
+            pred = tf.cast(pred, dtype=tf.int32)
+            # pred = tf.cast(tf.argmax(logits, axis=1), tf.int32)
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/fedavg/algorithm/basemodel.py:171]: Annotated code should be removed

File: examples/cifar100/fci_ssl/glfc/algorithm/agumentation.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    def __call__(self, images):
+        return images
+
+
+class Weak_Augment(Base_Augment):
+    def __init__(self, dataset_name: str) -> None:
+        super().__init__(dataset_name)
+        if self.dataset_name in ['cifar10', 'cifar100']:
+            self.augment_impl = self.augment_for_cifar
+        elif self.dataset_name == 'svhn':
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/glfc/algorithm/agumentation.py:133]: Unused dataset augmentation code

File: examples/cifar100/fci_ssl/glfc/algorithm/basemodel.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    def __init__(self, **kwargs) -> None:
+        self.kwargs = kwargs
+        self.learning_rate = kwargs.get("learning_rate", 0.001)
+        self.epochs = kwargs.get("epochs", 1)
+        self.batch_size = kwargs.get("batch_size", 32)
+        self.task_size = kwargs.get("task_size", 10)
+        self.memory_size = kwargs.get("memory_size", 2000)
+        self.encode_model = lenet5(32, 100)
+        self.encode_model.call(keras.Input(shape=(32, 32, 3)))
+        # print(self.encode_model.get_weights())
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/glfc/algorithm/basemodel.py:41]: Annotated code should be removed

File: examples/cifar100/fci_ssl/glfc/algorithm/basemodel.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        print(self.encode_model.summary())
+        # keras.initializers.glorot_uniform()
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/glfc/algorithm/basemodel.py:43]: Annotated code should be removed

File: examples/cifar100/fci_ssl/glfc/algorithm/model.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import keras
+
+# import keras
+# from keras import layers, Sequential
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/glfc/algorithm/model.py:19]: Annotated code should be removed

File: examples/cifar100/fci_ssl/glfc/algorithm/proxy_server.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+                                gy = tf.cast(gy, tf.double)
+                                sub_value = tf.subtract(gx, gy)
+                                pow_value = tf.pow(sub_value, 2)
+                                grad_diff += tf.reduce_sum(pow_value)
+                            # print(f'grad_diff {grad_diff} and grad_diff shape {grad_diff.shape}')
+                        grad = tape0.gradient(grad_diff, dummy_data)
+                        # print(f' grad_diff shape {grad.shape} and type(grad) {type(grad)}')
+                        opt.apply_gradients(zip([grad], [dummy_data]))
+                        
+                        # if iter == self.Iteration - 1:
```
Comment (chenhaorui0768) [inline examples/cifar100/fci_ssl/glfc/algorithm/proxy_server.py:189]: Annotated code should be removed

File: examples/cifar100/federated_class_incremental_learning/fedavg/algorithm/basemodel.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        return result
+
+    def eval(self, data, round, **kwargs):
+        total_num = 0
+        total_correct = 0
+        data = self.data_process(data)
+        # print(f"in evalute data: {data}")
+        for i, (x, y) in enumerate(data):
+            logits = self.model(x, training=False)
+            # prob = tf.nn.softmax(logits, axis=1)
```
Comment (chenhaorui0768) [inline examples/cifar100/federated_class_incremental_learning/fedavg/algorithm/basemodel.py:145]: Annotated code should be removed

File: examples/cifar100/federated_class_incremental_learning/fedavg/algorithm/network.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+class NetWork(keras.Model):
+    def __init__(self, num_classes, feature_extractor):
+        super(NetWork, self).__init__()
+        self.num_classes = num_classes
+        self.feature = feature_extractor
+        self.fc = Dense(num_classes, activation='softmax')
+
+    def call(self, inputs):
+        # print(type(self.feature))
```
Comment (chenhaorui0768) [inline examples/cifar100/federated_class_incremental_learning/fedavg/algorithm/network.py:30]: Annotated code should be removed

File: examples/cifar100/federated_class_incremental_learning/fedavg/algorithm/resnet.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import keras
+# import keras
```
Comment (chenhaorui0768) [inline examples/cifar100/federated_class_incremental_learning/fedavg/algorithm/resnet.py:17]: Annotated code should be removed

File: examples/cifar100/utils.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar100.load_data()
+    print(y_test.shape)
+    # 数据预处理：归一化
+    # x_train, x_test = x_train / 255.0, x_test / 255.0
+
+    # 将标签转换为类别索引
+    class_labels = np.unique(y_train)  # 获取所有类别
+    train_class_dict = {label: [] for label in class_labels}
+    test_class_dict = {label: [] for label in class_labels}
+    # train_cnt = 0
```
Comment (chenhaorui0768) [inline examples/cifar100/utils.py:39]: Annotated code should be removed

**Review by chenhaorui0768 — CHANGES_REQUESTED** [review]

**Comment by MooreZheng** [comment]
@hsj576 might want to have a look at this project

**Review by hsj576 — CHANGES_REQUESTED** [review]
The EGG-INFO file should not be uploaded to the github repository.

**Comment by Yoda-wu** [comment]
> The EGG-INFO file should not be uploaded to the github repository.

Sorry about that， i remove it  right now

**Review by MooreZheng — COMMENTED** [review]
This pull request contains 8 commits, which might make maintenance difficult, considering the number of contributors, pull requests, and their commits in KubeEdge Ianvs recently. In the final stage, @Yoda-wu can squash the commits into one using [rebase techniques](https://blog.kinboy.wang/2019/04/09/Git-%E5%8E%8B%E7%BC%A9%E5%A4%9A%E4%B8%AAcommit%E4%B8%BA%E5%8D%95%E4%B8%AAcommit/).

**Comment by Yoda-wu** [comment]
> This pull request contains 8 commits, which might make maintenance difficult, considering the number of contributors, pull requests, and their commits in KubeEdge Ianvs recently. In the final stage, @Yoda-wu can squash the commits into one using [rebase techniques](https://blog.kinboy.wang/2019/04/09/Git-%E5%8E%8B%E7%BC%A9%E5%A4%9A%E4%B8%AAcommit%E4%B8%BA%E5%8D%95%E4%B8%AAcommit/).

thank you for the advice! i will squash the commits if this pr have nothing to change. Looking forward to more reviews from @MooreZheng @chenhaorui0768 @hsj576. Due to the deadline of OSPP, please feel free to comment so i can fix it asap~

**Review by MooreZheng — APPROVED** [review]
All concerns are fixed. Well done @Yoda-wu
